components/membership/membership-persistence-client-impl/src/main/kotlin/net/corda/membership/impl/persistence/client/AbstractPersistenceClient.kt

```
@@ -63,55 +54,26 @@ abstract class AbstractPersistenceClient(
-            return persistenceFailureResponse(failureReason)
-        }
-        logger.info("Sending membership persistence RPC request.")
-
-        return try {
-            val response = sender
-                .sendRequest(this)
-                .getOrThrow(Duration.ofMillis(RPC_TIMEOUT_MS))
+    fun <T> MembershipPersistenceRequest.operation(
+        convertResult: (Any?) -> Either<T, String>,
+    ): MembershipPersistenceOperationImpl<T> {
+        return MembershipPersistenceOperationImpl(
+            rpcSender,
+            this,
+            convertResult,
+        )
+    }
 
-            with(context) {
-                require(holdingIdentity == response.context.holdingIdentity) {
-                    "Holding identity in the response received does not match what was sent in the request."
-                }
-                require(requestTimestamp == response.context.requestTimestamp) {
-                    "Request timestamp in the response received does not match what was sent in the request."
-                }
-                require(requestId == response.context.requestId) {
-                    "Request ID in the response received does not match what was sent in the request."
-                }
-                require(requestTimestamp <= response.context.responseTimestamp) {
-                    "Response timestamp is before the request timestamp"
-                }
-            }
-            response
-        } catch (e: IllegalArgumentException) {
-            persistenceFailureResponse("Invalid response. ${e.message}")
-        } catch (e: TimeoutException) {
-            persistenceFailureResponse("Timeout waiting for response from membership persistence RPC request.")
-        } catch (e: Exception) {
-            persistenceFailureResponse("Exception occurred while sending RPC request. ${e.message}")
+    fun MembershipPersistenceRequest.execute(): Any? {
+        val result = this.operation {
+            Either.Left(it)
```

This seems to be inconsistent with the other `operation` invocations from within `MembershipPersistenceClientImpl`.. Are we ignoring error here?

---

No, both the invocations from `MembershipPersistenceClientImpl` and `MembershipQueryClientImpl` are calling this common execute and then convert the payload.

---

Looking at `MembershipPersistenceClientImpl`, the 3 operations that return operations (persistRegistrationRequest, setMemberAndRegistrationRequestAsDeclined, setRegistrationRequestStatus) perform error handling as part of the `operation` invocation. The user code then invokes `execute` on the operation.

In contrast, all the other methods in `MembershipPersistenceClientImpl` invoke `MembershipPersistenceRequest.execute()`, which internally invokes `operation` without any error handling as it converts everything to `Left`. 

One example is if the rpc handler throws a `PersistenceFailedResponse`. In the first case, this will be handled, but in the second case I don't think it will be, because the `convertResult` lambda will convert everything to `Left`.

---

All of them have a when that captures the `PersistenceFailedResponse` after the execute. I wanted to reduce the change, but maybe consistency will be better.

---

right, I think I followed now.. the exception is indeed caught later.

But, it's a bit confusing, because the semantics of `convertResult` are it should return `Left` for successful results and `Right` for failures. And I believe what is happening here is `MembershipPersistenceRequest.execute()` will always return a `Left`. It's just that this is handled at a level above after `execute`.

It would be easier to follow if `operation` / `execute` methods have consistent semantics between `MembershipPersistenceOperationImpl` and `MembershipPersistenceRequest`.

---

It is a little confusing having the different `execute` functions with different semantics. It took me some time of looking over it to follow fully. I think though if the query and persistence clients were updated so that only persistence operations were returned it would be less of an issue since only the execute operation and create async commands functions would be exposed and I think `MembershipPersistenceRequest.execute()` could be internal then.

---

I had another look at this and I am not sure that converting return types for `MembershipQueryClientImp` to `MembershipPersistenceOperationImpl` might not make a lot of sense, since these are supposed to be synchronous queries in contrast to the methods from `MembershipPersistenceClientImpl`, where updates might need to be performed asynchronously. I see we made all the methods consistent in `MembershipPersistenceClientImpl` at least. So, could be ok for now.